enhancements/network/network-policy-audit-logging.md

```diff
@@ -0,0 +1,357 @@
+the ability to specify logging for each "allow" or "deny" rule.  This enhancement will activate the OVN ACL feature logging and allow the customer to manipulate the logging level, rate, and namespaces in which it is used, thereby showing valuable realtime information involving network policies.
+
+## Motivation
+
+Many customers require the ability to audit network policy related traffic events for regulatory and security policy compliance. Openshift currently does not have any features that satisfy this requirement. The ACL audit logging will allow customers to monitor NetworkPolicy events and identify patterns involving malicious activity with both allow and deny events. This is necessary in scenarios where
+customers require a certain level of compliance, such as monitoring firewall activity, intrusion detection support, or to perform post-mortem analysis.
+
+### Goals
+
+- Activate configurable ACL allow/deny logging in OVN-Kubernetes on a per namespace basis.
+- Allow the cluster administrator set the global logging via the Cluster Network Operator's configuration.
+- Allow the cluster administrator set the logging level via a namespace's `.metadata.annotations` field.
+- Collect the relevant data from the ovn-controller logs and present it to the cluster administrator.  
+
+### Non-Goals
+
+- Network Policy Object Logging, `oc describe <entity>` takes care of that.
+
+## Proposal
+
+To begin implementing ACL audit logging, first the [network.operator.openshift.io](https://github.com/openshift/api/blob/master/operator/v1/types_network.go) API needs to be updated.  These changes involve adding the optional `aclLoggingRateLimit` flag to the `OVNKubernetesConfig` struct as follows.
+
+```go
+// ovnKubernetesConfig contains the configuration parameters for networks
+// using the ovn-kubernetes network project
+type OVNKubernetesConfig struct {
+ // mtu is the MTU to use for the tunnel interface. This must be 100
+ // bytes smaller than the uplink mtu.
+ // Default is 1400
+ // +kubebuilder:validation:Minimum=0
+ // +optional
+ MTU *uint32 `json:"mtu,omitempty"`
+ // geneve port is the UDP port to be used by geneve encapulation.
+ // Default is 6081
+ // +kubebuilder:validation:Minimum=1
+ // +optional
+ GenevePort *uint32 `json:"genevePort,omitempty"`
+ // HybridOverlayConfig configures an additional overlay network for peers that are
+ // not using OVN.
+ // +optional
+ HybridOverlayConfig *HybridOverlayConfig `json:"hybridOverlayConfig,omitempty"`
+ // ipsecConfig enables and configures IPsec for pods on the pod network within the
+ // cluster.
+ // +optional
+ IPsecConfig *IPsecConfig `json:"ipsecConfig,omitempty"`
+ 
+ <--- BEGIN NEW CODE --->
+ // aclLoggingRateLimit allows for the configuring of max acl logging rate
```

Comment:
I strongly suspect we're going to need to add more auditing features. I'd suggest moving this "down a level" by adding an intermediate struct.

Something like
```go
type OVNKubernetesConfig struct {

    // PolicyAuditing is the configuration for network 
    PolicyAuditing *PolicyAuditConfig
}

type PolicyAuditConfig {

    // RateLimit is the approximate maximum number of messages to send per-second per-node. If
    // unset, no limit is applied.
    RateLimit *uint32

    // Destination is the destination for policy log messages. Messages are output in syslog format
    // Valid values are:
    // - "stdout" or "stderr: the audit logging container outputs to its stdout or stderr
    // - "udp:host:port" for sending syslog over UDP
    // - "tcp:host:port" for sending syslog over TCP
    Destination string

    // SyslogFacility the RFC5424 facility for generated messages, e.g. "kern". Default is "local0"
    SyslogFacility *string

    // DefaultLogLevelAllow is the syslog severity used for logging allow events.
    // Allow events are generated whenever a packet is accepted by a network policy, subject
    // to rate limiting. If unset, "allow" messages are not sent. This can also be configured on a
    // per-namespace basis.
    // acceptable values are emer, err, warn, info, or dbg
    DefaultLogLevelAllow *string

    // DefaultLogLevelDeny is the syslog severity used for logging deny events.
    // Allow events are generated whenever a packet is accepted by a network policy, subject
    // to rate limiting. If unset, "deny" messages are not sent. This can also be configured on a
    // per-namespace basis.
    // acceptable values are emer, err, warn, info, or dbg
    DefaultLogLevelDeny *LogLevel
}
```

**Note:** There is a mismatch between ovn severity and syslog severity. Maybe we can clean that up?

Reply:
We may also wish to enable ACL logging directly to the host's journal if there is rate limiting applied. @smarterclayton tends to have concerns about journald throughput, but in this case, we can cap the load.

Reply:
Agreed I really like this idea, would we allow host journal logging in parallel with logging to another destination? Or can we just have one of the destinations specified as `host`? (And if the `host` destination is chosen some default rate limit is applied)

Reply:
The only other note here is that I think for now the `DefaultLogLevelAllow` and `DefaultLogLevel` should not be included here since currently the CNO has no control over OVN-K's internal ACL logging levels... However this could be something we decide to implement in the future.